### PR TITLE
No `forall ->` in data constructors

### PR DESCRIPTION
**Pull request by goldfirere**
This strikes a small piece of #81 that was never implemented and deserves more careful study than this proposal gives it.

### CODE REVIEW

**Comment by monoidal** [comment]
Is this still active, considering #402?

**Comment by goldfirere** [comment]
I think so, in that we're moving this idea to a different proposal.

@nomeata, does this need full committee action, or can we just pull the trigger? (I myself am split on this.) In any case, I'd like to move this forward.

**Comment by simonpj** [comment]
I think we can just pull the trigger.  It's very much a dark corner.   Any objections from anyone?

**Comment by int-index** [comment]
My problem with this amendment is that it explicitly disallows `forall ->` in GADT constructors, while #402 allows it. So I think they are mutually exclusive. I think we should merge this amendment if #402 is rejected, but discard the amendment if #402 is accepted.

**Comment by simonpj** [comment]
This amendment is for #81, which is about kinds.  It says

> This new construct will be rejected in any context that is unambiguously a type for a term. 

It previously had an ill-thought-out bit about data constructors that are used only in types (in promoted form). This edit conservatively removes that bit.

Nothing stops us adding `forall a -> ` to the types of terms and data constructors later.  Until we do that, #402 might still be what we want but we should take the `forall a-> ` part out until we do.

TL;DR: I see no conflict here.

**Comment by int-index** [comment]
But a GADT constructor signature is not "unambiguously a type for a term", as when the constructor is promoted, the signature specifies its kind, where `forall ->` is allowed, today.

> ill-thought-out bit about data constructors 

It was not ill-thought-out, it was insufficiently thought out, and #402 provides the missing bits (nested foralls in constructor signatures)

**Comment by nomeata** [comment]
> @nomeata, does this need full committee action, or can we just pull the trigger? (I myself am split on this.) In any case, I'd like to move this forward.

I can’t tell. Simon PJ says go ahead, so maybe do? :-)

**Comment by simonpj** [comment]
For this particular edit, I think we can just go ahead.
